tools/ci/scripts/cve/osquery/github_api.py

```diff
@@ -0,0 +1,186 @@
+
+    def makePostRequest(self, url: str, data: dict):
+        time_passed = (time.time_ns() - self.last_request_time) / (1000 * 1000)
+
+        # We should in theory detect if the answer contains a Retry-After,
+        # because then that should be the delay to be used.
+        # We are not because currently this is only used for creating issues,
+        # and the Retry-After header is not sent in that case.
+        if time_passed < 5000:
+            sleep((5000 - time_passed) / 1000.0)
+
+        self.last_request_time = time.time_ns()
+
+        response = requests.post(
+            url,
+            json=data,
+            headers={
+                "Accept": "application/vnd.github.v3+json",
+                "Authorization": "token %s" % self.github_token_,
+            },
+        )
+
+        return response
+
+    def makeGetRequest(self, url, params: dict):
+        time_passed = (time.time_ns() - self.last_request_time) / (1000 * 1000)
+
+        # Limit at 1 request every 5 seconds
+        if time_passed < 5000:
+            sleep((5000 - time_passed) / 1000.0)
+
+        self.last_request_time = time.time_ns()
+
+        response = requests.get(
+            url,
+            params=params,
+            auth=("Bearer", self.github_token_),
+            headers={
+                "Accept": "application/vnd.github.v3+json",
+            },
+        )
+
+        return response
+
+    def createIssue(self, title: str, content: str, labels: list):
+
+        data = {
+            "title": title,
+            "body": content,
+            "labels": labels,
+        }
+
+        attempts = 0
+        while attempts < 3:
+            attempts += 1
+
+            self.debugPrint(f"Trying to open issue with title \"{data['title']}\"")
+
+            response = self.makePostRequest(self.dest_repo_url, data)
+
+            if response.status_code != 201:
+                self.debugPrint(
+                    f"Request to {self.dest_repo_url} to create issue with title"
+                    f" {data['title']} failed with {response.status_code}, reason:"
+                    f" {response.reason + ' ' + response.text if response.text else response.reason}"
+                )
+                sleep(attempts * 5)
+                continue
+
+            return response
+
+        raise IssueCreationError(
+            "Failed to open issue with status code: %s and reason: %s"
+            % (response.status_code, response.reason)
+        )
+
+    def getRecentOpenIssues(
+        self, creator="github-actions[bot]", state=GithubIssueState.Open
+    ) -> "list[list[dict]]":
+        # Query only issues that are open and have been created or updated in the last 6 months.
+        # Issues older than that are most likely not interesting
+        # because hopefully a release fixing them
+        # or something that ignores the CVEs have been created.
+        since = datetime.utcnow() - timedelta(days=180)
+
+        params = {
```

Comment:
I can't tell whether or not this is correctly searching, or leaving it for client side. If I read the python right, it looks like it's being submitted?

Either way, does `https://api.github.com/search/issues?q=is:issue%20repo:osquery/osquery%20label:security` get closer?

Reply:
The `since` is passed to the API call just below, in the `params` dictionary.

About the time limit, I partially misremembered during the office hours but, the first reason is indeed a limit on the issues being retrieved, but only on the CVE issues themselves. Either way, as you can see I can only retrieve 100 issues every 5 seconds. Hopefully it won't grow that quickly?

The second convenience that having a separate JSON file that provides the ignored CVEs, is that it partially stores a state, a point in time, that can be checked at any time.
It might not be as useful, but suppose the current version of osquery is 5.9.0 and I'm downloading the source code for 5.7.0 and wanted to check how was the state of the CVEs at that time, I could've re-run the script and had the view. You can also check the state down to the specific commit.

The issues instead always show the last state. That been said, I'm not saying it's a vital feature, just thought it was useful. We can emulate that in other ways I guess, by creating a list of the CVEs found, ignored and resolved that we report somewhere.

Reply:
The idea of having a snapshot in the release is a good one.

If we felt strongly about using Issues as a tracking DB, we could capture the state into a release file. 

I'm not totally sure where I fall.

What you have is mostly simple to implement. But I'm a little worried that needing to update the manifest will be cumbersome. But, the counter point is that updating dependancies is also cumbersome. So maybe that's fine. 

Where are you envisioning discussion about a CVE goes? Into the github issue?

Reply:
Just to compare and contrast a little more:

So, the Issues way is a bit more automatic, meaning, one closes a CVE and it's automatically regarded as ignored, without review. 
On the contrary the manifest way is less automatic, but it means that a review happened before closing an issue; if an issue is closed and no review happened, it will be reopened, much like not resolving a CVE we should not ignore, because no library update has been done.

Obviously as mentioned in the office hours, someone may close the issue without doing the PR, but at that point, opening a new issue with the same CVE is a signal that the procedure has not been followed. I could maybe even list closed CVEs and find one to reopen, if there is, instead of opening a duplicate.
I feel like deciding to ignore a CVE is a pretty important decision, so having a PR process with it was additionally my way of ensuring that there were enough eyes on it.

Finally, another issue comes to my mind (which I thought about when developing but did not immediately recorded). Hopefully not but the script may have bugs, or we want to change the format of the title; ignoring closed and old CVE issues was a way to permit change to happen there.
I felt like changing a JSON file was much quicker than going back to change all the old issues, or start adding support for different ways to parse the issue titles.

> Where are you envisioning discussion about a CVE goes? Into the github issue?
Ideally yes; as a "forum" and to keep track of the discussion I think that's best. We can obviously further discuss things on Slack, but the issues should at least capture a clear reasoning of why a CVE issue has been ignored/closed.